plugin.rb

```diff
@@ -26,44 +29,87 @@ class Engine < ::Rails::Engine
-      raise Discourse::InvalidParameters.new(:post_id) if post.blank?
-      guardian.ensure_can_see!(post)
-
       begin
-        detected_lang, translation = "DiscourseTranslator::#{SiteSetting.translator}".constantize.translate(post)
+        detected_lang, translation = DiscourseTranslator.current_service.translate(@object)
         render json: { translation: translation, detected_lang: detected_lang }, status: 200
       rescue ::DiscourseTranslator::TranslatorError => e
         render_json_error e.message, status: 422
       end
     end
+
+    protected
+
+    def ensure_plugin_enabled
+      raise PluginDisabled if !SiteSetting.translator_enabled
+    end
+
+    def rate_limit
+      if !current_user.staff?
+        RateLimiter.new(current_user, "translate", SiteSetting.max_translations_per_minute, 1.minute).performed!
+      end
+    end
+
+    def find_object
+      raise Discourse::InvalidParameters.new unless params[:post_id] || params[:topic_id]
+      @object = params[:post_id] ? Post.find_by(id: params[:post_id]) : Topic.find_by(id: params[:topic_id])
+      raise Discourse::InvalidParameters.new if @object.blank?
+      guardian.ensure_can_see!(@object)
+    end
   end
 
   Post.register_custom_field_type(::DiscourseTranslator::TRANSLATED_CUSTOM_FIELD, :json)
+  Topic.register_custom_field_type(::DiscourseTranslator::TRANSLATED_CUSTOM_FIELD, :json)
+
+  add_model_callback(:post, :before_update) do
+    if raw_changed?
+      custom_fields.delete(DiscourseTranslator::DETECTED_LANG_CUSTOM_FIELD)
+      custom_fields.delete(DiscourseTranslator::TRANSLATED_CUSTOM_FIELD)
+    end
+  end
+
+  add_model_callback(:topic, :before_update) do
+    if title_changed?
+      custom_fields.delete(DiscourseTranslator::DETECTED_TITLE_LANG_CUSTOM_FIELD)
+      custom_fields.delete(DiscourseTranslator::TRANSLATED_CUSTOM_FIELD)
+    end
+  end
+
+  add_to_class(:topic, :title_translations) do
+    translations = if !custom_fields[DiscourseTranslator::TRANSLATED_CUSTOM_FIELD].nil?
+      custom_fields[DiscourseTranslator::TRANSLATED_CUSTOM_FIELD].merge
+    else
+      {}
+    end
+  end
 
-  class ::Post < ActiveRecord::Base
-    before_update :clear_translator_custom_fields, if: :raw_changed?
+  add_to_class(:topic, :title_language) { custom_fields[DiscourseTranslator::DETECTED_TITLE_LANG_CUSTOM_FIELD] }
 
-    private
+  add_to_class(:topic, :translated_title) do
+    language = DiscourseTranslator.current_service::SUPPORTED_LANG_MAPPING[I18n.locale]
+    translation = title_translations[language]
+    translation ? translation : title
+  end
 
-    def clear_translator_custom_fields
-      return if !SiteSetting.translator_enabled
+  add_to_class(:topic, :translated_fancy_title) do
+    self.class.fancy_title(translated_title)
+  end
 
-      self.custom_fields.delete(DiscourseTranslator::DETECTED_LANG_CUSTOM_FIELD)
-      self.custom_fields.delete(DiscourseTranslator::TRANSLATED_CUSTOM_FIELD)
+  on(:post_created) do |post, options, user|
```

Comment:
This should only run if `post.is_first_post?` otherwise it will send translation requests for the title on every post.

Reply:
🤦‍♂️ fixed: https://github.com/angusmcleod/discourse-translator/commit/a4385a790eb25ad3e3d8a17e8cf17df02a66ccf5